server/src/main/java/org/elasticsearch/index/mapper/Mapper.java

```diff
@@ -22,6 +24,57 @@
 
 public abstract class Mapper implements ToXContentFragment, Iterable<Mapper> {
 
+    public static final NodeFeature STORE_SOURCE_MAPPER_PARAM = new NodeFeature("mapper.source.store_param");
+
+    static final String STORE_SOURCE_PARAM = "store_source";
+
+    // Only relevant for synthetic source mode.
+    public enum StoreSourceMode {
+        NONE("none"),      // No source recording
+        ARRAYS("arrays"),  // Store source for arrays of mapped fields
+        ALL("all");        // Store source for both singletons and arrays of mapped fields
+
+        StoreSourceMode(String name) {
+            this.name = name;
+        }
+
+        static StoreSourceMode from(String input) {
+            if (input.equals(NONE.name)) {
+                return NONE;
+            }
+            if (input.equals(ALL.name)) {
+                return ALL;
+            }
+            if (input.equals(ARRAYS.name)) {
+                return ARRAYS;
+            }
+            throw new IllegalArgumentException("Unknown " + STORE_SOURCE_PARAM + " value [" + input + "]");
+        }
+
+        @Override
+        public String toString() {
+            return name;
+        }
+
+        private final String name;
+    }
+
+    // Only relevant for indexes configured with synthetic source mode. Otherwise, it has no effect.
+    // Controls the default behavior for storing the source of leaf fields and objects, in singleton or array form.
+    // Setting to StoreSourceMode.ALL is equivalent to disabling synthetic source, so this is not allowed.
+    public static final Setting<StoreSourceMode> STORE_ARRAY_SOURCE_SETTING = Setting.enumSetting(
+        StoreSourceMode.class,
+        "index.mapping.store_source",
```

Comment:
To me, this same suggests that this is about storing source or related to the `_source` field. That isn't the case here and this setting only applies when synthetic source is active and controls whether arrays should be stored in its original form/structure. 

Maybe we can name this setting: `index.mapping.synthetic_source.keep` or `...retain`, which accepts the values allowed here?

Reply:
How about `index.mapping.synthetic_source.store_source`? This way, we keep the same (leaf) name as the per-field param that will be separately introduced - as @lkts called out too.

Reply:
If the attribute name is just `store_source` then it still is kind of confusing that this is a synthetic source only option? What about `synthetic_source_keep=none|arrays|all`?

Reply:
Done.

Reply:
`index.mapping.sysnthetic_source.preserve_array`?